code/game/objects/machinery/computer/dropship_weapons.dm

```diff
@@ -30,20 +30,135 @@
+
+			switch((selected_eqp_ammo_amt / selected_eqp_max_ammo_amt) * 100)
+				if(50 to INFINITY)
+					ammo_status = "good"
+				if(0 to 50)
+					ammo_status = "average"
+				if(-INFINITY to 0)
+					ammo_status = "bad"
+		screen_mode = selected_equipment.screen_mode	
+
+	data = list(	
+		"shuttle_mode" = shuttle.mode,	
+		"equipment_data" = equipment_data,	
+		"targets_data" = targets_data,	
+		"selected_eqp" = selected_eqp_name,	
+		"selected_eqp_ammo_name" = selected_eqp_ammo_name,	
+		"selected_eqp_ammo_amt" = selected_eqp_ammo_amt,	
+		"selected_eqp_max_ammo_amt" = selected_eqp_max_ammo_amt,	
+		"selected_eqp_ammo_status" = ammo_status,
+		"screen_mode" = screen_mode,	
+	)	
+	
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)	
+	if (!ui)	
+		ui = new(user, src, ui_key, "dropship_weapons_console.tmpl", "Weapons Control", 500, 400)	
+		ui.set_initial_data(data)	
+		ui.open()	
+		ui.set_auto_update(1)
+
+/obj/machinery/computer/dropship_weapons/Topic(href, href_list)	
+	. = ..()
+	if(.)	
+		return	
+
+	var/obj/docking_port/mobile/marine_dropship/shuttle = SSshuttle.getShuttle(shuttle_tag)
+	if(!shuttle)
+		stack_trace("Invalid shuttle_tag [shuttle_tag]")
+		return
+
+	if(href_list["equip_interact"])	
+		var/base_tag = text2num(href_list["equip_interact"])	
+		var/obj/structure/dropship_equipment/E = shuttle_equipments[base_tag]	
+		E.linked_console = src	
+		E.equipment_interact(usr)	
+
+	if(href_list["open_fire"])	
+		var/targ_id = text2num(href_list["open_fire"])	
+		var/mob/living/L = usr	
+		if(!istype(L))
+			return
+		if(L.mind.assigned_role != "Pilot Officer") //everyone can fire dropship weapons while fumbling.	
+			L.visible_message("<span class='notice'>[L] fumbles around figuring out how to use the automated targeting system.</span>",	
+			"<span class='notice'>You fumble around figuring out how to use the automated targeting system.</span>")	
+			var/fumbling_time = 100 - 20 * L.mind.cm_skills.pilot	
+			if(!do_after(L, fumbling_time, TRUE, src, BUSY_ICON_UNSKILLED))	
+				return FALSE	
+		for(var/X in GLOB.active_laser_targets)	
+			var/obj/effect/overlay/temp/laser_target/LT = X	
+			if(LT.target_id == targ_id)	
+				if(shuttle.mode != SHUTTLE_CALL)	
+					to_chat(L, "<span class='warning'>Dropship can only fire while in flight.</span>")	
+					return	
+				if(shuttle.mode == SHUTTLE_HIJACK_LOCK) 
+					return	
+
+				if(!selected_equipment || !selected_equipment.is_weapon)	
+					to_chat(L, "<span class='warning'>No weapon selected.</span>")	
+					return	
+				var/obj/structure/dropship_equipment/weapon/DEW = selected_equipment	
+				if(!DEW.ammo_equipped || DEW.ammo_equipped.ammo_count <= 0)	
+					to_chat(L, "<span class='warning'>[DEW] has no ammo.</span>")	
+					return	
+				if(DEW.last_fired > world.time - DEW.firing_delay)	
+					to_chat(L, "<span class='warning'>[DEW] just fired, wait for it to cool down.</span>")	
+					return	
+				if(!LT.loc) 
```

Comment:
It should alway have a loc, perhaps a QDELETED check? Or is this one necessary in some other way?